components/playlist/browser/resources/reducers/player_reducer.ts

```diff
@@ -73,6 +147,35 @@ const playerReducer: Reducer<PlayerState | undefined> = (
         }
       }
       break
+
+    case types.PLAYER_TOGGLE_AUTO_PLAY: {
+      state = { ...state, autoPlayEnabled: !state.autoPlayEnabled }
+      break
+    }
+
+    case types.PLAYER_TOGGLE_SHUFFLE: {
+      const shuffleEnabled = !state.shuffleEnabled
+      if (state.currentList) {
+        if (shuffleEnabled) {
+          const shuffledItems = shuffleItems(
+            [...state.currentList.items],
+            state.currentItem
+          )
+
+          state.itemsInOrder = [...state.currentList.items]
+          state.currentList = { ...state.currentList, items: shuffledItems }
+        } else {
+          state.currentList = {
+            ...state.currentList,
+            items: state.itemsInOrder!
+          }
+          state.itemsInOrder = undefined
```

Comment:
I wonder if in unshuffled mode we should just set `itemsInOrder: [...state.currentList.items]`, so we don't have to worry about null/undefined checks?